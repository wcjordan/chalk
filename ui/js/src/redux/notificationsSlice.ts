import { createSlice } from '@reduxjs/toolkit';

import { NotificationsState } from './types';

const initialState: NotificationsState = {
  notificationQueue: [],
};
export default createSlice({
  name: 'notifications',
  initialState,
  reducers: {
    addNotification: (state, action) => {
      state.notificationQueue.push(action.payload);
    },
    dismissNotification: (state) => {
      state.notificationQueue.shift();
    },
  },
});
