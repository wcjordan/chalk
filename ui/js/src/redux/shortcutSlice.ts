import { createSlice } from '@reduxjs/toolkit';

import { ShortcutState } from './types';

const initialState: ShortcutState = {
  operations: [],
  latestGeneration: 0,
};
export default createSlice({
  name: 'shortcuts',
  initialState,
  reducers: {
    addEditTodoOperation: (state, action) => {
      state.operations.push({
        type: 'EDIT_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    addMoveTodoOperation: (state, action) => {
      state.operations.push({
        type: 'MOVE_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    incrementGenerations: (state) => {
      state.latestGeneration = state.latestGeneration + 1;
    },
    clearOperationsUpThroughGeneration: (state, action) => {
      state.operations = state.operations.filter(
        (op) => op.generation > action.payload,
      );
    },
  },
});
