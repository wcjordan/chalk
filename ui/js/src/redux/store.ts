import { configureStore } from '@reduxjs/toolkit';
import rootReducer, { listTodos } from './reducers';

const store = configureStore({
  reducer: rootReducer,
});

// TODO dispose
window.setInterval(() => store.dispatch(listTodos()), 1000);

export default store;
