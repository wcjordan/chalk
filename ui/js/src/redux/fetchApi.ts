import Cookies from 'js-cookie';
import { Platform } from 'react-native';

import { getEnvFlags } from '../helpers';
import { RootState } from './store';

const webCsrfToken = Cookies.get('csrftoken') as string;
export function getCsrfToken(getState: () => RootState): string {
  if (Platform.OS === 'web') {
    return webCsrfToken;
  }
  return getState().workspace.csrfToken as string;
}

export async function list<T>(apiUri: string): Promise<Array<T>> {
  const response = await fetch(apiUri, getRequestOpts('GET'));
  return handleResponse<Array<T>>(response);
}

export async function patch<T, Q extends { id: number }>(
  apiUri: string,
  entryPatch: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('PATCH', csrfToken);
  requestOpts.body = JSON.stringify(entryPatch);
  const response = await fetch(`${apiUri}${entryPatch.id}/`, requestOpts);
  return handleResponse<T>(response);
}

export async function create<T, Q>(
  apiUri: string,
  newEntry: Q,
  csrfToken: string,
): Promise<T> {
  return postRequest<T, Q>(apiUri, newEntry, csrfToken);
}

export async function postRequest<T, Q>(
  uri: string,
  data: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = JSON.stringify(data);
  const response = await fetch(uri, requestOpts);
  return handleResponse<T>(response);
}

function getRequestOpts(
  method: string,
  csrfToken: string | null = null,
): RequestInit {
  const headers: HeadersInit = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
  if (csrfToken) {
    headers['X-CSRFToken'] = csrfToken;
  }

  const credentials: RequestCredentials = Platform.select({
    native: 'include',
    default: 'same-origin',
  });

  return {
    credentials: credentials,
    headers: headers,
    method: method,
  };
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(`${response.status} ${response.statusText}`);
  }
  return await response.json();
}

export function getWsRoot(): string {
  if (Platform.OS === 'web') {
    return '';
  }

  const subdomain =
    getEnvFlags().ENVIRONMENT === 'prod' ? 'chalk' : 'chalk-dev';
  return `http://${subdomain}.flipperkid.com/`
}

// Used to exchange login token for session cookie in mobile login flow
export async function completeAuthCallback(token: string) {
  return await fetch(
    `${getWsRoot()}api/todos/auth_callback/?code=${token}`,
    getRequestOpts('GET'),
  );
}

export async function logSessionData(
  dataString: string,
  csrfToken: string,
): Promise<string> {
  
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = dataString;
  const response = await fetch(
    `${getWsRoot()}api/todos/log_session_data/`,
    requestOpts,
  );
  return handleResponse<string>(response);
}
