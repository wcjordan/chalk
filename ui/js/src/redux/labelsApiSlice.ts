import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ApiState, ReduxState, Label } from './types';
import { getWsRoot, list } from './fetchApi';

const API_NAME = 'labelsApi';

// Exported for testing
export function getLabelsApi(): string {
  return `${getWsRoot()}api/todos/labels/`;
}

const initialState: ApiState<Label> = {
  entries: [],
  initialLoad: true,
  loading: false,
};

export const listLabels = createAsyncThunk<
  Label[],
  void,
  { state: ReduxState }
>(`${API_NAME}/list`, async () => list<Label>(getLabelsApi()), {
  condition: (_unused, { getState }) => {
    return !getState().labelsApi.loading;
  },
});

export default createSlice({
  name: API_NAME,
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(listLabels.pending, (state) => {
        state.loading = true;
      })
      .addCase(listLabels.fulfilled, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        state.entries = action.payload;
      })
      .addCase(listLabels.rejected, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        console.warn(`Loading Labels failed. ${action.error.message}`);
      });
  },
});
