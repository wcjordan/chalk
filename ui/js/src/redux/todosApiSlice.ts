import _ from 'lodash';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { Platform } from 'react-native';
import { ApiState, NewTodo, ReduxState, Todo, TodoPatch } from './types';
import { create, list, patch } from './fetchApi';

const API_NAME = 'todosApi';

function getWsRoot() {
  const wsroot = Platform.select({
    native: 'http://chalk-dev.flipperkid.com/',
    default: '',
  });
  return wsroot;
}

// Exported for testing
export function getTodosApi(): string {
  return getWsRoot() + 'api/todos/todos/';
}

interface ErrorAction {
  error: Error;
}

interface TodoAction {
  payload: Todo;
}

interface TodoListAction {
  payload: Array<Todo>;
}

const initialState: ApiState<Todo> = {
  entries: [],
  loading: false,
};

export const createTodo = createAsyncThunk<
  Todo,
  string,
  Record<string, unknown>
>(`${API_NAME}/create`, async (todoTitle: string) => {
  const newTodo = {
    created_at: Date.now(),
    description: todoTitle,
  };
  return create<Todo, NewTodo>(getTodosApi(), newTodo);
});

export const listTodos = createAsyncThunk<Todo[], void, { state: ReduxState }>(
  `${API_NAME}/list`,
  async () => list<Todo>(getTodosApi()),
  {
    condition: (_unused, { getState }) => {
      return !getState().todosApi.loading;
    },
  },
);

export const updateTodo = createAsyncThunk<
  Todo,
  TodoPatch,
  Record<string, unknown>
>(`${API_NAME}/update`, async (entryPatch) =>
  patch<Todo, TodoPatch>(getTodosApi(), entryPatch),
);

/**
 * Filter out archived todos and sort todos by id.
 * Also sort completed todos to appear at the bottom.
 */
function processTodos(todos: Array<Todo>) {
  return _.sortBy(
    _.filter(todos, (todo) => !todo.archived),
    ['completed', 'id'],
  );
}

export default createSlice({
  name: API_NAME,
  initialState,
  reducers: {},
  extraReducers: {
    [createTodo.fulfilled.type]: (state, action: TodoAction) => {
      state.entries.push(action.payload);
      state.entries = processTodos(state.entries);
    },
    [createTodo.rejected.type]: (_state, action: ErrorAction) => {
      console.warn(`Creating Todo failed. ${action.error.message}`);
    },
    [listTodos.pending.type]: (state) => {
      state.loading = true;
    },
    [listTodos.fulfilled.type]: (state, action: TodoListAction) => {
      state.loading = false;
      state.entries = processTodos(action.payload);
    },
    [listTodos.rejected.type]: (state, action: ErrorAction) => {
      state.loading = false;
      console.warn(`Loading Todo failed. ${action.error.message}`);
    },
    [updateTodo.fulfilled.type]: (state, action: TodoAction) => {
      const updatedEntry = action.payload;
      const existingEntry = _.find(
        state.entries,
        (entry) => entry.id === updatedEntry.id,
      );
      Object.assign(existingEntry, updatedEntry);
      state.entries = processTodos(state.entries);
    },
    [updateTodo.rejected.type]: (_state, action: ErrorAction) => {
      console.warn(`Updating Todo failed. ${action.error.message}`);
    },
  },
});
