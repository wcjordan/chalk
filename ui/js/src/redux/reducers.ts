import { ThunkAction } from 'redux-thunk';
import { Action } from '@reduxjs/toolkit';

import { ReduxState, TodoPatch } from './types';
import labelsApiSlice, { listLabels } from './labelsApiSlice';
import notificationsSlice from './notificationsSlice';
import shortcutSlice from './shortcutSlice';
import todosApiSlice, {
  createTodo,
  listTodos,
  updateTodo as updateTodoApi,
} from './todosApiSlice';
import workspaceSlice from './workspaceSlice';
import { completeAuthCallback } from './fetchApi';

type AppThunk = ThunkAction<void, ReduxState, unknown, Action<string>>;

export const updateTodo =
  (todoPatch: TodoPatch): AppThunk =>
  (dispatch) => {
    return Promise.all([
      dispatch(
        notificationsSlice.actions.addNotification(
          `Saving Todo: ${todoPatch.description}`,
        ),
      ),
      dispatch(workspaceSlice.actions.setEditTodoId(null)),
      dispatch(shortcutSlice.actions.addEditTodoOperation(todoPatch)),
      dispatch(updateTodoApi(todoPatch)),
    ]);
  };

export const updateTodoLabels =
  (labels: string[]): AppThunk =>
  (dispatch, getState) => {
    const todoId = getState().workspace.labelTodoId;
    if (todoId === null) {
      throw new Error('Unable to edit a todo w/ null ID');
    }

    return dispatch(
      updateTodoApi({
        id: todoId,
        labels,
      }),
    );
  };

// Used to exchange login token for session cookie in mobile login flow
export const completeAuthentication =
  (accessToken: string): AppThunk =>
  async (dispatch) => {
    try {
      const response = await completeAuthCallback(accessToken);
      const cookies = response.headers.get('set-cookie')?.split(' ');
      const csrfToken = cookies
        ?.find((val) => val.startsWith('csrftoken'))
        ?.slice(10, -1);

      if (response.status !== 200) {
        const responseText = await response.text();
        const message = `Login failed with status: ${response.status} ${response.statusText}\n${responseText}`;
        dispatch(notificationsSlice.actions.addNotification(message));
        throw new Error(message);
      }

      if (!csrfToken) {
        dispatch(
          notificationsSlice.actions.addNotification(
            'Unexpectedly missing CSRFToken.  Please refresh and login again.',
          ),
        );
        throw new Error(
          `CSRFToken missing from set-cookie header\n${response.headers.get(
            'set-cookie',
          )}`,
        );
      }

      dispatch(
        workspaceSlice.actions.logIn({
          loggedIn: true,
          csrfToken: csrfToken,
        }),
      );
    } catch (ex) {
      dispatch(
        notificationsSlice.actions.addNotification(
          'Login failed with a network error.  Please refresh and login again.',
        ),
      );
      throw ex;
    }
  };

export const addNotification = notificationsSlice.actions.addNotification;
export const dismissNotification =
  notificationsSlice.actions.dismissNotification;
export const setEditTodoId = workspaceSlice.actions.setEditTodoId;
export const setLabelTodoId = workspaceSlice.actions.setLabelTodoId;
export const setWorkContext = workspaceSlice.actions.setWorkContext;
export const toggleLabel = workspaceSlice.actions.toggleLabel;
export const toggleShowCompletedTodos =
  workspaceSlice.actions.toggleShowCompletedTodos;
export const toggleShowLabelFilter =
  workspaceSlice.actions.toggleShowLabelFilter;
export { createTodo, listLabels, listTodos };
export default {
  labelsApi: labelsApiSlice.reducer,
  notifications: notificationsSlice.reducer,
  shortcuts: shortcutSlice.reducer,
  todosApi: todosApiSlice.reducer,
  workspace: workspaceSlice.reducer,
};
