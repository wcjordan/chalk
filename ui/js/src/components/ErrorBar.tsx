import React, { useCallback, useState } from 'react';
import { Snackbar } from 'react-native-paper';
import { dismissNotification } from '../redux/reducers';
import { useAppDispatch } from '../hooks';

const ErrorBar: React.FC<Props> = function (props: Props) {
  const { permanent, text } = props;
  const [visible, setVisible] = useState(text != null);

  const dispatch = useAppDispatch();
  const dismissCb = useCallback(() => {
    if (text == null || permanent) {
      return;
    }

    setVisible(false);
    dispatch(dismissNotification());
  }, [text]);

  return (
    <Snackbar
      onDismiss={dismissCb}
      theme={{
        colors: {
          surface: '#444',
          onSurface: '#FAA0A0',
        },
      }}
      visible={visible}
    >
      {text}
    </Snackbar>
  );
};

type Props = {
  permanent?: boolean;
  text: string | null;
};

export default ErrorBar;
