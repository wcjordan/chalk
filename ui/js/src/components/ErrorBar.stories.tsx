import React from 'react';
import { StyleSheet, View } from 'react-native';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { Provider } from 'react-redux';
import { setupStore } from '../redux/store';
import ErrorBar from './ErrorBar';

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    height: '100%',
    paddingTop: '60px',
    padddingBottom: '40px',
  },
});
const wrapper = (component) => (
  <SafeAreaProvider>
    <Provider store={setupStore()}>
      <View style={styles.wrapper}>{component}</View>
    </Provider>
  </SafeAreaProvider>
);

export default {
  title: 'Error Bar',
  component: ErrorBar,
};
export const DefaultErrorBar: React.FC = () => wrapper(<ErrorBar text="Snacks!" />);

export const HiddenErrorBar: React.FC = () => wrapper(<ErrorBar text={null} />);

export const PermanentErrorBar: React.FC = () =>
  wrapper(<ErrorBar text="Permanent snacks!" permanent={true} />);
