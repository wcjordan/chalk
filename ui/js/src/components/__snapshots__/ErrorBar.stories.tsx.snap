// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Bar DefaultErrorBar smoke-test 1`] = `
<div class="css-g5y9jx r-13awgt0">
  <div class="css-g5y9jx r-18u37iz r-1w6e6rj r-1pi2tsx r-10q4ly9 r-u04ygo">
    <div class="css-g5y9jx r-1p0dtai r-u8s1d r-13qz1uu r-12vffkv"
         style="padding-bottom: 0px; padding-right: 0px; padding-left: 0px;"
    >
      <div aria-live="polite"
           class="css-g5y9jx r-18u37iz r-1wtj0ep r-hwh8t1 r-1pl7oy7 r-12vffkv"
           data-testid="surface"
           style="box-shadow: rgba(0, 0, 0, 0.3) 0px 2px 6px; background-color: rgb(49, 48, 51); border-radius: 4px; opacity: 1; transform: scale(1);"
      >
        <div dir="auto"
             class="css-146c3p1 r-fdjqy7 r-13awgt0 r-1apnidv r-1xpp3t0"
             style="direction: ltr; font-family: Roboto, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; letter-spacing: 0.25px; font-weight: 400; line-height: 20px; font-size: 14px; color: rgb(244, 239, 244);"
        >
          Snacks!
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Error Bar HiddenErrorBar smoke-test 1`] = `
<div class="css-g5y9jx r-13awgt0">
  <div class="css-g5y9jx r-18u37iz r-1w6e6rj r-1pi2tsx r-10q4ly9 r-u04ygo">
  </div>
</div>
`;

exports[`Error Bar PermanentErrorBar smoke-test 1`] = `
<div class="css-g5y9jx r-13awgt0">
  <div class="css-g5y9jx r-18u37iz r-1w6e6rj r-1pi2tsx r-10q4ly9 r-u04ygo">
    <div class="css-g5y9jx r-1p0dtai r-u8s1d r-13qz1uu r-12vffkv"
         style="padding-bottom: 0px; padding-right: 0px; padding-left: 0px;"
    >
      <div aria-live="polite"
           class="css-g5y9jx r-18u37iz r-1wtj0ep r-hwh8t1 r-1pl7oy7 r-12vffkv"
           data-testid="surface"
           style="box-shadow: rgba(0, 0, 0, 0.3) 0px 2px 6px; background-color: rgb(49, 48, 51); border-radius: 4px; opacity: 1; transform: scale(1);"
      >
        <div dir="auto"
             class="css-146c3p1 r-fdjqy7 r-13awgt0 r-1apnidv r-1xpp3t0"
             style="direction: ltr; font-family: Roboto, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; letter-spacing: 0.25px; font-weight: 400; line-height: 20px; font-size: 14px; color: rgb(244, 239, 244);"
        >
          Permanent snacks!
        </div>
      </div>
    </div>
  </div>
</div>
`;
